apps/prairielearn/assets/scripts/instructorAssessmentAccessClient.ts

```diff
@@ -0,0 +1,216 @@
+  const accessRulesTable = document.querySelector('.js-access-rules-table');
+  const editAccessRuleModal = document.querySelector('#editAccessRuleModal');
+  const deleteAccessRuleModal = document.querySelector('.js-delete-access-rule-modal');
+  const addRuleButton = document.getElementById('addRuleButton');
+
+  const accessRulesData = decodeData('access-rules-data');
+
+  const ptHost = (accessRulesTable as HTMLElement)?.dataset.ptHost ?? '';
+  const devMode = (accessRulesTable as HTMLElement)?.dataset.devMode === 'true';
+  const hasCourseInstancePermissionView =
+    (accessRulesTable as HTMLElement)?.dataset.hasCourseInstancePermissionView === 'true';
+  const timezone = (accessRulesTable as HTMLElement)?.dataset.timezone ?? 'UTC';
+
+  let editMode = false;
+
+  function refreshTable() {
+    morphdom(
+      accessRulesTable as Node,
+      AccessRulesTable({
+        accessRules: accessRulesData,
+        ptHost,
+        devMode,
+        hasCourseInstancePermissionView,
+        editMode,
+        timezone,
+      }).toString(),
+    );
+  }
+
+  enableEditButton?.addEventListener('click', () => {
+    editMode = true;
+    enableEditButton.style.display = 'none';
+    editModeButtons?.style.removeProperty('display');
+    addRuleButton?.style.removeProperty('display');
+    refreshTable();
+  });
+
+  on('click', '#saveAndSyncButton', () => {
+    const form = document.getElementById('accessRulesForm') as HTMLFormElement;
+    const assessmentAccessRulesInput = form.querySelector(
+      'input[name="assessment_access_rules"]',
+    ) as HTMLInputElement;
+    if (!assessmentAccessRulesInput) return;
+    const accessRulesMap = accessRulesData.map((accessRule: Record<string, any>) => {
+      const rule = {
+        mode: accessRule.assessment_access_rule.mode,
+        uids: accessRule.assessment_access_rule.uids
+          ? accessRule.assessment_access_rule.uids
+          : null,
+        startDate: adjustedDate(accessRule.assessment_access_rule.start_date)
+          .toISOString()
+          .slice(0, 19),
+        endDate: adjustedDate(accessRule.assessment_access_rule.end_date)
+          .toISOString()
+          .slice(0, 19),
+        active: accessRule.assessment_access_rule.active,
+        credit: parseInt(accessRule.assessment_access_rule.credit),
+        timeLimitMin: parseInt(accessRule.assessment_access_rule.time_limit_mins),
+        password: accessRule.assessment_access_rule.password,
+        examUuid: accessRule.assessment_access_rule.exam_uuid,
+      };
+      const filteredRules = Object.fromEntries(Object.entries(rule).filter(([_, value]) => value));
+      return filteredRules;
+    });
+    assessmentAccessRulesInput.value = JSON.stringify(accessRulesMap);
+    form.submit();
+  });
+
+  function adjustedDate(dateString: string | Date) {
+    const date = new Date(dateString);
+    const timezoneOffset = date.getTimezoneOffset() * 60000;
+    return new Date(date.getTime() - timezoneOffset);
+  }
```

Comment:
My hope was that this function wouldn't be necessary after the work we did this afternoon. Is that not the case? If it is indeed still necessary, it'd be nice to see if we can do it with the `Temporal` API, and we should document what this is doing and why.

Reply:
I have tried doing this with Temporal and still can't seem to get it to work. Of course, you have seen me try to use the `Temporal` API and continue to get turned around. If there is a way to do this with `Temporal` I would love to be able to use that instead. However, I have instead continued using this function for the time being and added a JSDoc comment to document what it is doing.